File: backend-python/routes/activities.py
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from lib.supabase_client import supabase
from middleware.auth import verify_token, verify_admin

router = APIRouter()


class ActivityBody(BaseModel):
    activity_type: str
    description: str = None
    metadata: dict = {}


@router.post("/", status_code=201)
async def log_activity(body: ActivityBody, user: dict = Depends(verify_token)):
    res = (
        supabase.from_("activities")
        .insert([{
            "user_id": None if user["id"] == "default-admin" else user["id"],
            "username": user["username"],
            "activity_type": body.activity_type,
            "description": body.description,
            "metadata": body.metadata,
        }])
        .select()
        .single()
        .execute()
    )
    return {"activity": res.data}


@router.get("/")
async def get_activities(
    limit: int = Query(50, ge=1, le=200),
    user: dict = Depends(verify_admin),
):
    res = (
        supabase.from_("activities")
        .select("*")
        .order("created_at", desc=True)
        .limit(limit)
        .execute()
    )
    return {"activities": res.data}


@router.get("/my")
async def my_activities(user: dict = Depends(verify_token)):
    res = (
        supabase.from_("activities")
        .select("*")
        .eq("user_id", user["id"])
        .order("created_at", desc=True)
        .limit(20)
        .execute()
    )
    return {"activities": res.data}


@router.get("/stats")
async def activity_stats(user: dict = Depends(verify_admin)):
    res = supabase.from_("activities").select("activity_type").execute()
    stats = {
        "total": len(res.data),
        "login": 0, "admin_login": 0, "market_search": 0,
        "prediction": 0, "register": 0, "logout": 0,
    }
    for a in res.data:
        if a["activity_type"] in stats:
            stats[a["activity_type"]] += 1
    return {"stats": stats}

File: backend/routes/activities.js
const express = require('express');
const router = express.Router();
const supabase = require('../lib/supabase');
const { verifyToken, verifyAdmin } = require('../middleware/auth');

// POST /api/activities  - Log an activity
router.post('/', verifyToken, async (req, res) => {
  try {
    const { activity_type, description, metadata } = req.body;
    if (!activity_type) return res.status(400).json({ error: 'activity_type is required' });

    const { data, error } = await supabase.from('activities').insert([{
      user_id: req.user.id === 'default-admin' ? null : req.user.id,
      username: req.user.username,
      activity_type,
      description,
      metadata: metadata || {}
    }]).select().single();

    if (error) throw error;
    res.status(201).json({ activity: data });
  } catch (err) {
    console.error('Log activity error:', err);
    res.status(500).json({ error: 'Failed to log activity' });
  }
});

// GET /api/activities  - Get all activities (admin only)
router.get('/', verifyAdmin, async (req, res) => {
  try {
    const limit = parseInt(req.query.limit) || 50;
    const { data, error } = await supabase
      .from('activities')
      .select('*')
      .order('created_at', { ascending: false })
      .limit(limit);
    if (error) throw error;
    res.json({ activities: data });
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch activities' });
  }
});

// GET /api/activities/my  - Get current user's activities
router.get('/my', verifyToken, async (req, res) => {
  try {
    const { data, error } = await supabase
      .from('activities')
      .select('*')
      .eq('user_id', req.user.id)
      .order('created_at', { ascending: false })
      .limit(20);
    if (error) throw error;
    res.json({ activities: data });
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch activities' });
  }
});

// GET /api/activities/stats  - Activity stats for dashboard
router.get('/stats', verifyAdmin, async (req, res) => {
  try {
    const { data, error } = await supabase.from('activities').select('activity_type');
    if (error) throw error;
    const stats = { total: data.length, login: 0, admin_login: 0, market_search: 0, prediction: 0, register: 0, logout: 0 };
    data.forEach(a => { if (stats[a.activity_type] !== undefined) stats[a.activity_type]++; });
    res.json({ stats });
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch stats' });
  }
});

module.exports = router;
